Reduce thread contention by shuffling rfile order

ddanielr opened the pull request:
Scans can hang on the MultiIterator.seek method due to iterators waiting for the block cache lock for a given rfile. 
Shuffling the rfile opening order should reduce the frequency of this cache locking problem.

ddanielr commented:
Need to add test cases.

dlmarion commented on server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/ScanDataSource.java:
```diff
@@ -177,8 +177,10 @@ private SortedKeyValueIterator<Key,Value> createIterator()
+    List<InterruptibleIterator> mapfiles =
         fileManager.openFiles(files, scanParams.isIsolated(), samplerConfig);
+    // Randomize the ordering of files to avoid block cache contention on seeks
+    Collections.shuffle(mapfiles);
```
I think you can use `tablet.getTableConfiguration()` to see if a shuffle property is set on the table. Another option, which is not currently wired up, would be to use an execution hint on a per-scan basis so that not all scans on the table are affected.

ddanielr commented:
Worked through this with @keith-turner and found that the shuffling could be done lower in the MultiIterator before things are added to the priorityQueue. 

Realized we had to do it in two places because of deepCopy. I added a test for deepCopy since that didn't exist. 

I'm not sure if we need to gate this change behind a property or not so I've removed that code for now. 
The MultiIterator is a system iterator so it's not part of the public API

ddanielr commented:
Ran the sunny ITs locally and this passed.

dlmarion reviewed (commented):
I'm not sure we should make this change in a patch release without making it configurable and defaulting to `off`. I think this thread contention issue - where you have multiple scans hitting the same tablet at the same time resulting in the same set of files in the same order - likely only happens in a few cases:

     - a table with very few tablets
     - a case where you have a thundering herd of like queries starting at the same time

keith-turner commented on core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/MultiIterator.java:
```diff
@@ -50,6 +51,7 @@ private MultiIterator(MultiIterator other, IteratorEnvironment env) {
     super(other.iters.size());
     this.iters = new ArrayList<>();
     this.fence = other.fence;
+    Collections.shuffle(other.iters);
```
These shuffles will make seeks happen in different order.  There is still the case of initially opening the RFiles in FileManager that we could also shuffle.

ddanielr replied:
I believe I handled the FileManager case in bb6102a
